### ISSUE TITLE
Protecting API endpoint

### ISSUE THREAD

**Issue by connorbenton** [authelia/authelia#395]
Hi, I'm a longtime user and Authelia has been working flawlessly on all of my self-hosted sites. I'm working on a personal project and it was very trivial to protect the frontend with Authelia, and now I'm curious if there's any way I'm also able to protect the backend through the auth_verify flow as well. Currently, all backend requests from the frontend site are just getting the standard 302 response (location: https://login.xxx.xxx/#/?rd=https://api.xxx.xxx/posts) - is there any way with the NGINX proxy settings or other methods to pass/proxy pass the authelia_session cookie (that I already have in all my requests to the frontend) in these API requests so that Authelia sees that I'm an authorized user and allows my requests through to the API backend? Thanks!

**Comment by clems4ever** [comment]
Hello @connorbenton , thank you for the feedback on your usage of Authelia. That's nice to see it's used by so many people now.
Regarding your question, everything depends on the configuration you wrote for nginx. If Authelia is supposed to protect the API as well as the frontend, the validation of the session should be automatic (given you have the right permissions in Authelia config). Otherwise, there might be a missing part in the configuration of nginx letting the session cookie be forwarded to Authelia.

Paste your config so that we can help you out.

**Comment by connorbenton** [comment]
Thanks! This is the specific nginx block I have - I have a wildcard LetsEncrypt cert for everything under *.XXX.cloud

`server {
listen  80;
server_name     XXXapi.XXX.cloud;
return  301 https://$server_name$request_uri;
}

server {

server_name XXXapi.XXX.cloud;
listen 443 ssl http2;
listen [::]:443 ssl http2;
include /etc/nginx/snippets/strong-ssl.conf;
ssl_certificate /etc/letsencrypt/live/XXX.cloud/fullchain.pem;
ssl_certificate_key /etc/letsencrypt/live/XXX.cloud/privkey.pem;
set $upstream_endpoint http://127.0.0.1:8080;
set $upstream_verify http://127.0.0.1:8080/api/verify;

client_max_body_size 10M;
add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
add_header X-Frame-Options "SAMEORIGIN";

# Change the client side error pages (4xx) to prevent some information disclosure
error_page 401 403 404 /404.html;

include authelia_check-auth_block_internal_api;

location /auth_verify {
internal;
proxy_set_header  Host $http_host;
proxy_set_header  X-Original-URI $request_uri;
proxy_set_header  X-Original-URL $scheme://$http_host$request_uri;
proxy_set_header  X-Real-IP $remote_addr;
proxy_set_header  X-Forwarded-Proto $scheme;


proxy_pass_request_body     off;
proxy_set_header  Content-Length "";
proxy_pass  $upstream_verify;
}

location / {
auth_request  /auth_verify;
auth_request_set $user $upstream_http_remote_user;
auth_request_set $groups $upstream_http_remote_groups;
auth_request_set $target_url $scheme://$http_host$request_uri;

error_page 401 =302 https://login.XXX.cloud/#/?rd=$target_url;

proxy_set_header 'Access-Control-Allow-Origin' '*';
proxy_set_header 'Access-Control-Allow_Credentials' 'true';
proxy_set_header 'Access-Control-Allow-Methods' 'GET,POST,OPTIONS,PUT,DELETE,PATCH';

proxy_pass http://localhost:9029;

    proxy_http_version 1.1;
    proxy_set_header   Host    $http_host;
    proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header   Upgrade $http_upgrade;
    proxy_set_header   Connection      "Upgrade";
}
}`

the check-auth_block_internal contains:

`
location = /.check-auth {
    internal;
    proxy_set_header Host $host;
    proxy_set_header X-Original-URI $request_uri;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Content-Length "";
    proxy_pass_request_body off;
    proxy_pass http://127.0.0.1:8080/api/verify;
}`

**Comment by clems4ever** [comment]
Just a quick question, are you sure the `authelia_session` cookie is forwarded correctly with your requests targeting the backend API?
You should be able to see it in Authelia logs if you're in debug mode.

**Comment by connorbenton** [comment]
This is what I have - doesn't look like the session ID is successfully
passing, could it possibly be a CORS issue within Authelia?

{"log":"debug: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='Headers =
{\"host\":\"XXXapi.XXX.cloud\",\"x-original-uri\":\"/api/saltEdgeConnections\",\"x-original-url\":\"
https://XXXapi.XXX.cloud/api/saltEdgeConne
ctions\",\"x-real-ip\":\"192.168.2.87\",\"x-forwarded-proto\":\"https\",\"connection\":\"close\",\"accept\":\"application/json,
text/plain, */*\",\"user-agent\":\"Mozilla/5.0 (X11; Linux x86_
64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90
Safari/537.36\",\"sec-fetch-mode\":\"cors\",\"origin\":\"
https://XXX.XXX.cloud\",\"sec-fetch-site\":\"same-site\",\"referer
\":\"https://XXX.XXX.cloud/accounts\",\"accept-encoding\":\"gzip, deflate,
br\",\"accept-language\":\"en-US,en;q=0.9\"}'\n","stream":"stderr","time":"2019-09-28T10:25:44.978347076Z"}
{"log":"debug: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='Authentication session
X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S was undefined. Resetting... If it's
unexpected, make sure you are visiting the expected domain.'\n","strea
m":"stderr","time":"2019-09-28T10:25:44.978384416Z"}
{"log":"debug: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='Header x-original-url is set to
https://XXXapi.XXX.cloud/api/saltEdgeConnections
'\n","stream":"stderr","time":"2019-09-28T10:25:44.978482802Z"}
{"log":"debug: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='domain=XXXapi.XXX.cloud,
path=/api/saltEdgeConnections, user=unknown, groups=unknown,
ip=::ffff:127.0.0.1'\n","stream":"stderr","time":"2019-09-28T10:25:44.9
78507722Z"}
{"log":"debug: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='Header x-original-url is set to
https://XXXapi.XXX.cloud/api/saltEdgeConnections
'\n","stream":"stderr","time":"2019-09-28T10:25:44.97867313Z"}
{"log":"error: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='Reply with error 401: User 'unknown' is not
sufficiently authorized to access
XXXapi.XXX.cloud/api/saltEdgeConnections.'\n","stream":"stderr","time":"2019-09
-28T10:25:44.978682795Z"}
{"log":"debug: date='Sat Sep 28 2019 10:25:44 GMT+0000 (UTC)' method='GET',
path='/api/verify' requestId='05c03c60-3aac-4896-8eea-c3c6e1b782e4'
sessionId='X4jOm09yK93FI2o7EWO7BG0zJ4P8DR4S' ip
='::ffff:127.0.0.1' message='NotAuthenticatedError: User 'unknown' is not
sufficiently authorized to access
XXXapi.XXX.cloud/api/saltEdgeConnections.\n","stream":"stderr","time":"2019-09
-28T10:25:44.978794879Z"}
{"log":"    at Object.default_1 [as default]
(/usr/src/server/src/lib/routes/verify/CheckAuthorizations.js:31:15)\n","stream":"stderr","time":"2019-09-28T10:25:44.978807606Z"}
{"log":"    at Object.\u003canonymous\u003e
(/usr/src/server/src/lib/routes/verify/GetSessionCookie.js:29:38)\n","stream":"stderr","time":"2019-09-28T10:25:44.978812087Z"}
{"log":"    at Generator.next
(\u003canonymous\u003e)\n","stream":"stderr","time":"2019-09-28T10:25:44.97881635Z"}
{"log":"    at
/usr/src/server/src/lib/routes/verify/GetSessionCookie.js:7:71\n","stream":"stderr","time":"2019-09-28T10:25:44.978819987Z"}
{"log":"    at new Promise
(\u003canonymous\u003e)\n","stream":"stderr","time":"2019-09-28T10:25:44.978884418Z"}
{"log":"    at __awaiter
(/usr/src/server/src/lib/routes/verify/GetSessionCookie.js:3:12)\n","stream":"stderr","time":"2019-09-28T10:25:44.9788927Z"}
{"log":"    at Object.default_1 [as default]
(/usr/src/server/src/lib/routes/verify/GetSessionCookie.js:17:12)\n","stream":"stderr","time":"2019-09-28T10:25:44.978896573Z"}
{"log":"    at
/usr/src/server/src/lib/routes/verify/Get.js:25:45\n","stream":"stderr","time":"2019-09-28T10:25:44.978899994Z"}
{"log":"    at Generator.next
(\u003canonymous\u003e)\n","stream":"stderr","time":"2019-09-28T10:25:44.978903268Z"}
{"log":"    at
/usr/src/server/src/lib/routes/verify/Get.js:7:71\n","stream":"stderr","time":"2019-09-28T10:25:44.978906517Z"}
{"log":"    at new Promise
(\u003canonymous\u003e)\n","stream":"stderr","time":"2019-09-28T10:25:44.97890998Z"}
{"log":"    at __awaiter
(/usr/src/server/src/lib/routes/verify/Get.js:3:12)\n","stream":"stderr","time":"2019-09-28T10:25:44.978913365Z"}
{"log":"    at verifyWithSelectedMethod
(/usr/src/server/src/lib/routes/verify/Get.js:20:12)\n","stream":"stderr","time":"2019-09-28T10:25:44.978916759Z"}
{"log":"    at
/usr/src/server/src/lib/routes/verify/Get.js:38:19\n","stream":"stderr","time":"2019-09-28T10:25:44.97892023Z"}
{"log":"    at Generator.next
(\u003canonymous\u003e)\n","stream":"stderr","time":"2019-09-28T10:25:44.978923556Z"}
{"log":"    at
/usr/src/server/src/lib/routes/verify/Get.js:7:71'\n","stream":"stderr","time":"2019-09-28T10:25:44.978926823Z"}

Thanks,
Connor

**Comment by clems4ever** [comment]
Hello @connorbenton , can you please try with Authelia v4?

**Comment by nightah** [comment]
@connorbenton I'm closing this issue off, please attempt with Authelia v4 and if you continue to have issues either post here with your configuration and logs so we can reopen this issue, or create a new issue.

**Comment by Janbong** [comment]
Hi, I realize I'm a bit late to re-open this issue, but  I'm facing a VERY similar problem here.

Trying to protect a Rest API (let's call it `api.example.com` from now on) with Authelia by letting request to it redirect to Authelia login first (if not logged in yet). If I already am logged in with Authelia, though, I can just access `api.exampIe.com` because I have the `authelia_session` cookie stored in my browser. Pretty basic up untill here, I guess.

The problem I am trying to solve is:
I want to access my Rest API from another web app (also protected by Authelia, on the same domain). Let's call this web app `webapp.example.com`. I'm trying to pass on the `authelia_session` cookie, which I have because I need to be authenticated with Authelia before being able to access `webapp.example.com`.
Passing on in this context means using the `authelia_session` from `webapp.example.com` to make REST calls to `api.example.com`.

I first encountered the problem when I tried to make a POST rest call to `api.example.com`. For some reason I don't understand, my Vue webapp (`webapp.example.com`) or Axios or a combination of both first makes an OPTIONS call and only after that executes the POST call. This is called CORS pre-flight and that's the only thing I know about it.
By adding a rule in Authelia that let's OPTIONS calls bypass authentication, I was able to make the OPTIONS and then POST calls. In the POST calls I can see that the `authelia_session` cookie is there and passed on. So I thought I had the whole thing fixed. Until I added a GET endpoint to my `api.example.com`. When I make a GET call from `webapp.example.com`, Axios/Vue/Webapp does not do CORS pre-flight and I get another CORS error. Adding GET to the bypass rule fixes it of course, but that defeats the purpose of protecting the API in the first place.

Can anyone help me?
